multimodal/src/autogluon/multimodal/data/labelencoder_ner.py

```diff
@@ -0,0 +1,175 @@
+
+from ..constants import AUTOMM, END_OFFSET, ENTITY_GROUP, NER_ANNOTATION, START_OFFSET
+from .utils import process_ner_annotations
+
+logger = logging.getLogger(AUTOMM)
+
+
+class NerLabelEncoder:
+    """
+    Label Encoder for the named entity recognition task.
+    """
+
+    def __init__(self, entity_map: Optional[dict] = None):
+        self.entity_map = entity_map
+
+    def fit(self, y: pd.Series, x: pd.Series, config: DictConfig):
+        """
+        Extract the annotations, check the unique entity groups, and build entity to index mappings.
+        """
+        model_config = config.model.ner
+        self.ner_special_tags = OmegaConf.to_object(model_config.special_tags)
+        self.prefix = config.model.names[0]
+        _, entity_groups = self.extract_ner_annotations(y)
+        self.unique_entity_groups = self.ner_special_tags + entity_groups
+        self.entity_map = {entity: index for index, entity in enumerate(self.unique_entity_groups)}
+        self.inverse_entity_map = {index: entity for index, entity in enumerate(self.unique_entity_groups)}
+        logger.debug(f"Unique entity groups in the data: {entity_groups}")
+
+    def extract_ner_annotations(self, y: pd.Series):
+        """
+        Validate the JSON annotations with predefined JSON schema and convert it into list.
+
+        Parameters
+        ----------
+        y
+            The raw json annotations.
+        Returns
+        -------
+        all_annotations
+            A list of NER annotations.
+        unique_entity_groups
+            A list of unique entity groups.
+        """
+
+        # Predefined Json schema
+        schema = {
+            "$schema": "http://json-schema.org/draft-04/schema#",
+            "type": "array",
+            "items": {
+                "type": "object",
+                "properties": {
+                    ENTITY_GROUP: {"type": "string"},
+                    START_OFFSET: {"type": "integer"},
+                    END_OFFSET: {"type": "integer"},
+                },
+                "required": [ENTITY_GROUP, START_OFFSET, END_OFFSET],
+            },
+        }
+        all_annotations = []
+        all_entity_groups = []
+        for ner_annotations in y.iteritems():
+            json_ner_annotations = json.loads(ner_annotations[-1])  # load the json annotations
+            try:
+                jsonschema.validate(json_ner_annotations, schema)  # verify the json schema
+            except jsonschema.ValidationError as e:
+                # Raise an error if the provided json annotations do not match the predefined schema
+                raise ValueError(f"The provided json annotations are invalid: {e.message}")
+            sentence_annotations = []
+            for annot in json_ner_annotations:
+                all_entity_groups.append(annot[ENTITY_GROUP])
+                if self.entity_map is not None:
+                    if annot[ENTITY_GROUP] in self.entity_map:
```

Comment:
`if annot[ENTITY_GROUP] not in self.entity_map`, then no action is taken. Is this expected?

Reply:
yes, it is expected. For example, if there is a new entity group (which is not seen in the training stage) in a validation/test datasets, the model will ignore it. Without this if statement, an error will be raised in such cases.